src/hooks/wallets/useProvider.ts

```diff
@@ -0,0 +1,34 @@
+import { useMemo } from 'react'
+import type { JsonRpcProvider, Web3Provider } from '@ethersproject/providers'
+
+import { useCurrentChain } from '@/hooks/useChains'
+import useWallet from '@/hooks/wallets/useWallet'
+import { createSafeAppsWeb3Provider, createWeb3, createWeb3ReadOnly } from '@/hooks/wallets/web3'
+import { useAppSelector } from '@/store'
+import { selectRpc } from '@/store/settingsSlice'
+
+export const useProvider = (isSafeApp = false): JsonRpcProvider | Web3Provider | undefined => {
+  const chain = useCurrentChain()
+  const wallet = useWallet()
+  const rpcMap = useAppSelector(selectRpc)
+
+  const customRpc = chain ? rpcMap[chain.chainId] : undefined
```

Comment:
This can be moved inside the `useMemo` and doesn't need to check for the existance of `chain` anymore.